include/dsn/dist/replication/client_ddl.h

```diff
@@ -0,0 +1,114 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/*
+ * Description:
+ *     ddl client interface
+ *
+ * Revision history:
+ *     2015-12-30, xiaotz, first version
+ */
+
+#include <cctype>
+#include <dsn/dist/replication.h>
+#include <dsn/dist/replication/replication.types.h>
+
+
+using namespace dsn::replication;
+
+namespace dsn{ namespace client{
+
+class client_ddl : public clientlet
+{
+public:
+    client_ddl(std::vector<dsn::rpc_address> meta_servers);
+
+    dsn::error_code create_app(const std::string& app_name, const std::string& app_type, int partition_count, int replica_count);
+
+    dsn::error_code drop_app(const std::string& app_name);
+
+private:
+    bool inline static valid_app_char(int c);
+
+    void inline end_meta_request(task_ptr callback, error_code err, dsn_message_t request, dsn_message_t resp);
+
+    template<typename TRequest, typename TResponse>
+    dsn::task_ptr request_meta(
+            dsn_task_code_t code,
+            std::shared_ptr<TRequest>& req,
+
+            // callback
+            clientlet* owner,
+            std::function<void(error_code, std::shared_ptr<TRequest>&, std::shared_ptr<TResponse>&)> callback,
+
+            // other specific parameters
+            int timeout_milliseconds= 0,
+            int reply_hash = 0
+            )
+    {
+        dsn_message_t msg = dsn_msg_create_request(code, timeout_milliseconds, 0);
+        ::marshall(msg, *req);
+
+        task_ptr task = ::dsn::rpc::internal_use_only::create_rpc_call(
+                 msg,
+                 req,
+                 callback,
+                 reply_hash,
+                 owner
+                 );
+        rpc_address target(_meta_servers);
+        rpc::call(
+            target,
+            msg,
+            this,
+            std::bind(&client_ddl::end_meta_request,
+            this,
+            task,
+            std::placeholders::_1,
+            std::placeholders::_2,
+            std::placeholders::_3
+            ),
+            0
+         );
+        return std::move(task);
+    }
+
+private:
+    dsn::rpc_address _meta_servers;
+};
+
+bool client_ddl::valid_app_char(int c)
+{
+    return (bool)std::isalnum(c) || c == '_';
+}
+
+void client_ddl::end_meta_request(task_ptr callback, error_code err, dsn_message_t request, dsn_message_t resp)
+{
+    callback->enqueue_rpc_response(err, resp);
```

Comment:
switch leader for _meta_server when err == timeout so that we have a chance to find the right leader in case the previous target meta server is gone.